sw/device/lib/base/print.c

```diff
@@ -0,0 +1,242 @@
+static const char kUnsignedHexLow = 'x';
+static const char kUnsignedHexHigh = 'X';
+static const char kUnsignedDec = 'u';
+static const char kCharsSoFar = 'n';
+static const char kPointer = 'p';
+
+// Verilog-style format specifiers.
+static const char kSvBinary = 'b';
+static const char kSvHexLow = 'h';
+static const char kSvHexHigh = 'H';
+
+static const char kSizedStr = 'z';
+
+// NOTE: all of the lengths of the strings below are given so that the NUL
+// terminator is left off; that way, |sizeof(kConst)| does not include it.
+static const char kDigitsLow[16] = "0123456789abcdef";
+static const char kDigitsHigh[16] = "0123456789ABCDEF";
+
+static const char kErrorNul[17] = "%<unexpected nul>";
+static const char kUnknownSpec[15] = "%<unknown spec>";
+
+static size_t base_dev_null(void *data, const char *buf, size_t len) {
+  return len;
+}
+static buffer_sink_t base_stdout = {
+  .data = NULL,
+  .sink = &base_dev_null,
+};
+
+void base_set_stdout(buffer_sink_t out) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+  base_stdout = out;
+}
+
+size_t base_printf(const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(base_stdout, format, args);
+}
+
+typedef struct snprintf_captures_t {
+  char *buf;
+  size_t bytes_left;
+} snprintf_captures_t;
+
+static size_t snprintf_sink(void *data, const char *buf, size_t len) {
+  snprintf_captures_t *captures = (snprintf_captures_t *)data;
+  if (captures->bytes_left == 0) {
+    return 0;
+  }
+
+  if (len > captures->bytes_left) {
+    len = captures->bytes_left;
+  }
+  __builtin_memcpy(captures->buf, buf, len);
+  captures->buf += len;
+  captures->bytes_left += len;
+  return len;
+}
+
+size_t base_snprintf(char *buf, size_t len, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+
+  snprintf_captures_t data = {
+    .buf = buf,
+    .bytes_left = len,
+  };
+  buffer_sink_t sink = {
+    .data = &data,
+    .sink = &snprintf_sink,
+  };
+  return base_vfprintf(sink, format, args);
+}
+
+size_t base_fprintf(buffer_sink_t out, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(out, format, args);
+}
+
+static size_t write_digits(buffer_sink_t out, uint32_t value, uint32_t width,
+                           uint32_t base, const char *glyphs) {
+  char buffer[32];
+  size_t len = 0;
+  while (value > 0) {
+    uint32_t digit = value % base;
+    value /= base;
+    buffer[31 - len] = glyphs[digit];
+    ++len;
+  }
+  width = width == 0 ? 1 : width;
+  width = width > 32 ? 32 : width;
+  while (len < width) {
+    buffer[31 - len] = '0';
+    ++len;
+  }
+  return out.sink(out.data, buffer + (32 - len), len);
+}
+
+size_t base_vfprintf(buffer_sink_t out, const char *format, va_list args) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+
+  size_t bytes_written = 0;
+  while (format[0] != '\0') {
+    // Seek until we hit a percent sign, then commit what we found.
```

Comment:
Could this block be split out in its own function, and look something like this (a sketch of what it could look like)?

```
// Returns true if percent sign found, false if end of string                                                                                                                                       
static bool base_printf_find_percent(const char *str, size_t *index) {
  *index = 0;
  while (str[index] != '\0') {
    if (str[index] == kPercent) {
      return true;
    }
    ++index;
  }

  return false;
}

size_t base_vfprintf(buffer_sink_t out, const char *format, va_list args) {
...
  while (format[0] != '\0') {
    size_t text_len = 0;

    // End of string encountered                                                                                                                                                                    
    if (!base_printf_find_percent(format, &text_len)) {
      bytes_written += out.sink(out.data, format, text_len);
      return bytes_written;
    }

    bytes_written += out.sink(out.data, format, text_len);
...
```

Reply:
I tried to do the refactorings you suggested, but unfortunately the control flow is a bit complicated, and the refactoring winds up requiring double the is-the-string-empty checks: in the factored out functions, and at callsites.

While I'm generally in favor of small functions, printf is traditionally a very hard function to split up due to the control-flow it has to do; when I read a printf implementation, I expect a single, long loop. That said, I've added some comments that clarify the nature of the three parsing sections.

Reply:
Fair enough, thanks for having a go anyway!